### PR TITLE
Merge turn_cacher into main repo

### PR DESCRIPTION
**Pull request by RLKRo**
# Checklist

- [x] [`functools.cache`](https://docs.python.org/3/library/functools.html#functools.cache) was only introduced in python3.9. Possible solution: replace it with `lru_cache(maxsize=None)`.
- [x] add examples

### CODE REVIEW

File: dff/core/engine/core/actor.py [hunk 1]
```diff
@@ -119,8 +123,6 @@ def __init__(
         condition_handler: Optional[Callable] = None,
         verbose: bool = True,
         handlers: Optional[Dict[ActorStage, List[Callable]]] = None,
-        *args,
```
Comment (kudep) [inline dff/core/engine/core/actor.py:122]: Revert it, how it was in `dev`

File: dff/script/utils/singleton_turn_caching.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+USED_CACHES = list()
+
+
+def cache_clear():
+    for used_cache in USED_CACHES:
+        used_cache.cache_clear()
+
+
+def lru_cache(maxsize: Optional[int] = None, typed: bool = False) -> Callable:
+    global USED_CACHES
```
Comment (kudep) [inline dff/script/utils/singleton_turn_caching.py:14]: Do we need use global two times in `lru_cache` and `decorator`?

File: dff/script/utils/singleton_turn_caching.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        def wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+
+        USED_CACHES += [wrapper]
+        return wrapper
+
+    return decorator
+
+
+def cache(func, /):
```
Comment (kudep) [inline dff/script/utils/singleton_turn_caching.py:29]: Should only `lru_cache` be kept instead `cache`? I think  we can drop `cache`

File: examples/cache/1_cache.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from dff.core.pipeline import Pipeline
+from dff.script.utils.singleton_turn_caching import cache
+from dff.utils.testing.common import check_happy_path, is_interactive_mode, run_interactive_mode
+
+
+external_data = {"counter": 0}
+
+
+@cache
+def cached_response(_):
```
Comment (kudep) [inline examples/cache/1_cache.py:14]: Can we add more comments, that for same `args` function return same for one turn. Need more comments and user guided docs.

**Review by kudep — CHANGES_REQUESTED** [review]
need to resolve issues

File: dff/script/utils/singleton_turn_caching.py (hunk 3 above)
Comment (pseusys) [reply dff/script/utils/singleton_turn_caching.py:29]: This one exists in standart lib, if we follow it's naming why shouldn't we follow it's function definitions?

File: dff/core/engine/core/actor.py [hunk 5]
```diff
@@ -120,7 +124,6 @@ def __init__(
         verbose: bool = True,
         handlers: Optional[Dict[ActorStage, List[Callable]]] = None,
         *args,
-        **kwargs,
```
Comment (kudep) [inline dff/core/engine/core/actor.py:123]: Revert it, how it was in `dev`

File: dff/script/utils/singleton_turn_caching.py (hunk 3 above)
Comment (kudep) [reply dff/script/utils/singleton_turn_caching.py]: Ok, I agree

**Review by kudep — CHANGES_REQUESTED** [review]
- [x] add description, that caching is erased in the end of the turn
- [x] document cache functions

File: dff/script/utils/singleton_turn_caching.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    Decorator function for caching function results in scripts.
+    Works like the standard `functools.lru_cache` function.
+    Caches are kept in a library-wide singleton and cleared in the end of each turn.
+    """
+
+    def decorator(func):
+        global USED_CACHES
+
+        @functools.lru_cache(maxsize=maxsize, typed=typed)
```
[the remark this answers is not in the document]
Comment (kudep) [reply dff/script/utils/singleton_turn_caching.py:29]: good idea

**Review by avsakharov — COMMENTED** [review]
Add new line

**Review by kudep — APPROVED** [review]

**Pull request merged**